### PR TITLE
Validating minimal required gas for a transaction

### LINKED ISSUE #2880: Verify base gas before adding transaction to the queue
Currently actual execution of transaction may fail:
`Execution(NotEnoughBaseGas { required: 21000, got: 5000 }))`

We could pre-verify minimal gas requirement before adding to the queue.

### PR DESCRIPTION
**Pull request by tomusdrw**
Closes #2880

### CODE REVIEW

**Comment by rphmeier** [comment]
LGTM. Might be useful to add a comment warning not to use any expensive checks for `gas_required` (e.g. executing the transaction to find how much it really costs) as that would be a DoS vector. Looks like this is fine, since it just checks the base gas requirement for `CALL`/`CREATE`?

File: ethcore/src/miner/banning_queue.rs [hunk 1]
```diff
@@ -75,11 +75,15 @@ impl BanningTransactionQueue {
...
-	pub fn add_with_banlist<F>(
+	pub fn add_with_banlist<F, G>(
 		&mut self,
 		transaction: SignedTransaction,
 		account_details: &F,
-	) -> Result<TransactionImportResult, Error> where F: Fn(&Address) -> AccountDetails {
+		gas_estimator: &G,
+	) -> Result<TransactionImportResult, Error>
+		where F: Fn(&Address) -> AccountDetails,
+			  G: Fn(&SignedTransaction) -> U256,
```
Comment (gavofyork) [inline ethcore/src/miner/banning_queue.rs]: no spaces mixing in with the tabs!!

**Comment by gavofyork** [comment]
check them spaces!

**Comment by tomusdrw** [comment]
Fixed

**Pull request merged**